---
PRACTICE CASE NoSQL
---
Pada project ini akan membuat collection baru dengan nama `clean_movies` yang sama persis dengan `movie_collection` pada database sample_mflix dengan memakai collection `movies_initial`.
Parameter validasi collection :
1. Semua document pada clean_movies dan movie sama 
2. Banyak document pada clean_movies dan movie sama 
3. Semua fields pada clean_movies ada pada movie 
4. Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

STEP 1: Import Library & Initiate Connection
------
> Langkah pertama yang dilakukan adalah **import package** dari modul pymonggo. Setelah dilakukan import modul, maka langkah selanjutnya adalah **initiate connection** sesuai dengan arahan yang telah ditentukan. 

In [2]:
### import MongoClient class from pymongo module
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

In [3]:
#### initiate connection 
client = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority")

STEP 2: Choose and Identification Database & Collections
------
> Langkah selanjutnya adalah mengetahui database yang ada pada client dan selanjutnya memilih collection yang digunakan. Database berisikan beberapa collection. Didalam collection terdapat document dimana dalam document tersebut berisikan fields dan value.

In [228]:
# Identifikasi Database pada Client
client.list_database_names()

['sample_mflix', 'admin', 'local']

Pada case ini, database yang digunakan adalah **sample_mflix**. Selanjutnya adalah mengetahui collection apa saja yang terdapat dalam database tersebut.

In [40]:
# Identifikasi Collection yang ada pada Database 
db = client['sample_mflix']
print(db.list_collection_names())

['clean_nama_kamu', 'clean_movies_nurlailiis', 'clean_movies_devita', 'movies_initial', 'clean_movies_frans', 'clean_movies_tara', 'clean_movies_tunggal', 'clean_movies', 'users', 'clean_movies_nurrimah', 'movies_testtopik', 'clean_movies_firdaus', 'movies_scratch', 'clean_movies_topik', 'movies', 'clean_movies_afifah', 'clean_movies_andreas', 'clean_movies_indra_1']


Didalam database sample_mflix terdapat beberapa collection, pada case ini collection yang digunakan adalah `movies` dan `movies_initial`.

In [38]:
db = client['sample_mflix']
col = db.movies

### Mengetahui key yang ada pada collection movies
x = col.find_one()
print(x.keys())

dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes'])


In [85]:
db = client['sample_mflix']
col1 = db.movies_initial

### Mengetahui key yang ada pada collection movies_initial
x1 = col1.find_one()
print(x1.keys())

dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type'])


Pada masing-masing collection memiliki jumlah keys yang berbeda sehingga ketika membuat collection baru yang berisikan `movies_initial` dimana format field menyesuaikan `movies` maka akan dibuat menggunakan **pipeline**.

In [231]:
#menghitung jumlah data di collection movies dan pengecekan format IMDB
imdb=list(col.find({},{'_id':0,'imdb':1}))
imdb_list=[]
print(len(imdb))

for t in imdb:
    imdb_list.append(list(t.values())[0])
imdb_list[:5]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

jumlah document yang terdapat di collection movies sebanyak 23541 document. Selain itu, untuk format dari IMDB diketahui berbentuk array.

In [248]:
#menghitung jumlah data di collection movies
imdb1=list(col1.find({},{'_id':0,'imdb':1}))
imdb_list1=[]
print(len(imdb1))

46014


Jumlah document pada collection `movies_initial` sebanyak 46.014.

STEP 3: Create New Collection
------
> Pada step ini, akan membuat collection baru dengan nama `clean_movies_devita`. Field yang berada pada collection tersebut berasal dari `movies_initial` dimana format field akan dibuat sama dengan `movies`.

In [215]:
#Drop collection clean_movies_devita untuk replace atau melakukan pembaharuan document pada collection
db.clean_movies_devita.drop()

pipeline = [
    {
        '$project':{
            'awards':{"$cond": [{"$eq": ["$awards", ""]}, "$nonExistinField", {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      }]} ,
            'cast' : {"$cond": [{"$eq": ["$cast", ""]}, "$nonExistinField", {'$split': ["$cast", ", "]}]},
            'countries': {"$cond": [{"$eq": ["$countries", ""]}, "$nonExistinField", {'$split': ["$country", ", "]}]},          
            'directors': {"$cond": [{"$eq": ["$director", ""]}, "$nonExistinField", {'$split': ["$director", ", "]}]},
            'fullplot': {"$cond": [{"$eq": ["$fullplot", ""]}, "$nonExistinField", "$fullplot"]},
            'genres': {"$cond": [{"$eq": ["$genre", ""]}, "$nonExistinField", {'$split': ["$genre", ", "]}]},
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },            
            'languages': {"$cond": [{"$eq": ["$language", ""]}, "$nonExistinField", "$language"]},
            'lastupdated': {"$cond": [{"$eq": ["$lastupdated", ""]}, "$nonExistinField", "$lastupdated"]},
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'num_mflix_comments': "",
            'plot': {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},
            'poster': {"$cond": [{"$eq": ["$poster", ""]}, "$nonExistinField", "$poster"]},
            'rated': {"$cond": [{"$eq": ["$rating", ""]}, "$nonExistinField", "$rating"]},
            'released': {"$cond": [{"$eq": ["$released", ""]}, "$nonExistinField",  {'$cond': {
                'if': {'$ne': ["$released", ""]},
                'then': { '$dateFromString': 
                         {
                            'dateString': "$released"
                        }
                    },
                    'else': ""}}]} , 
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, "$nonExistinField" ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]}, #100 min
            'title': {"$cond": [{"$eq": ["$title", ""]}, "$nonExistinField", "$title"]},          
            'tomatoes': {
                'lastUpdated': "$lastupdated",
                'viewer': {'numReviews': 1, 'rating': "$rating"},
                },
            'type': {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]},
            'writers': {"$cond": [{"$eq": ["$writer", ""]}, "$nonExistinField", {'$split': ["$writer", ", "]}]},
            'year': {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]}
        }
    },
    {
        '$match':{"imdb":{'$in':[dict(sorted(imdb_list[x].items(),key=lambda x: x[0])) for x in range(len(imdb_list))]}}
    },
    {
        ### Penamaan collection baru sebagai output
        '$out': "clean_movies_devita"
    }
]

clear_output ()
#melakukan agregasi yang telah telah didefinisikan menggunakan pipeline
pprint.pprint(list(db.movies_initial.aggregate(pipeline)))

[]


Setelah berhasil membentuk collection yang baru, dilakukan pengecekan terhadap hasil agregat yang telah dilakukan dengan menggunakan pipeline. 

In [216]:
list(db.clean_movies_devita.find({'title':"Titanic"}))

[{'_id': ObjectId('5db003c9d0a60ca4acd0cf20'),
  'awards': {'nominations': '2',
   'text': 'Won 1 Oscar. Another 2 nominations.',
   'wins': None},
  'cast': ['Clifton Webb',
   'Barbara Stanwyck',
   'Robert Wagner',
   'Audrey Dalton'],
  'countries': ['USA'],
  'directors': ['Jean Negulesco'],
  'fullplot': 'Unhappily married and uncomfortable with life among the British upper crust, Julia Sturges takes her two children and boards the Titanic for America. Her husband Richard also arranges passage on the doomed luxury liner in order to let him have custody of their two children. Their problems soon seem minor when the ship hits an iceberg.',
  'genres': ['Drama', 'History', 'Romance'],
  'imdb': {'id': 46435, 'rating': 7.3, 'votes': 4677},
  'languages': 'English, Basque, French, Spanish',
  'lastupdated': '2015-09-16 00:00:16.593000000',
  'num_mflix_comments': '',
  'plot': 'An unhappy married couple deal with their problems on board the ill-fated ship.',
  'poster': 'http://ia.med

STEP 3: Validation
------
> Setelah membuat collection baru, selanjutnya melakukan validasi terhadap collection baru dengan collection `movies`. Validasi yang dilakukan sebanyak 4 poin yang dilebur menjadi 3 poin untuk mengetahui apakah collection `clean_movies_devita` memiliki kesamaan dengan collection `movies`.

### Dokumen dan Banyak document pada clean_movies dan movie sama

In [219]:
#menghitung jumlah data di collection movies
len(list(col.find({},{'_id':0,'imdb':1})))

23541

In [220]:
#menghitung jumlah data di collection clean_movies_devita
len(list(db.clean_movies_devita.find({},{'_id':0,'imdb':1})))

23539

Terdapat perbedaan pada kedua collection, untuk mengetahui titik perbedaan diantara keduanya maka dilakukan identifikasi dari masing-masing value dari `movies` dan `clean_movies_devita`

In [221]:
grup = list(db.clean_movies_devita.aggregate([{"$sortByCount":"$imdb"}]))
grup2 = list(col.aggregate([{"$sortByCount":"$imdb"}]))
for i in grup2:
    if i not in grup:
        print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


Terdapat perbedaan pada imdb id: 5, rating: 6.2, votes: 1189 sebanyak 2. Selanjutnya akan dilakukan pengecekan terhadap kedua collection

In [222]:
#pengecekan pada collection clean_movies_devita
list(db.clean_movies_devita.find({"imdb.id":{"$in":[5]}}))

[{'_id': ObjectId('5db003a2d0a60ca4acd0bee5'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': '',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': '1',
  'title': 'Blacksmith Scene',
  'tomat

In [218]:
#pengecekan pada collection movies
list(col.find({"imdb.id":{"$in":[5]}}))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene',
  'tomatoes': {'

**Keterangan :** <br>

Setelah dilakukan pengecekan pada collection `clean_movies_devita` dan `movies` diketahui bahwa pada film **Blacksmith Scene II** terdapat tiga document di `movies` yang memiliki imdb id: 5, rating: 6.2, votes: 1189 dan hanya satu document pada `clean_movies_devita` yang memiliki imdb id: 5, rating: 6.2, votes: 1189. Sehingga diketahui selisih jumlah data diantara kedua collection adalah 2. Hal ini sesuai dengan perbedaan jumlah data diantara kedua collection tersebut yakni sebanyak 2 document.<br>

### Semua fields pada clean_movies ada pada movie 

Dalam melakukan validasi pada case ini, akan dilakukan pengecekan terhadap key yang ada pada masing-masing collection telah sama antar collection atau belum.

In [210]:
# Show all keys MOVIES 
b = [list(doc.keys()) for doc in col.find({})]
b=set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [224]:
# Show all keys Clean_Movies_Devita
a = [list(doc.keys()) for doc in db.clean_movies_devita.find({})]
a=set([j for i in a for j in i])
a

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

Dari kedua output diatas diketahui bahwa antara field yang ada pada collection `Movies` dan `clean_movies_devita` **tidak memiliki perbedaan antar field** Antara jumlah dan field yang ada di kedua collection memiliki kesamaan. 

### Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

Untuk mengetahui apakah value pada kedua collection sama, maka akan dilakukan pengecekan antar dua collection tersebut. Pada case ini, akan dilakukan pengecekan pada salah satu judul movie. 

In [247]:
pprint.pprint(col.find_one())

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
               'Thomas H. Ince',
               'Raymond B. West',
               'Walter Edwards',
               'David Hartford',
               'Jay Hunt',
               'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Natura

In [245]:
list(db.clean_movies_devita.find({'title':"Civilization"}))

[{'_id': ObjectId('5db003a2d0a60ca4acd0bf67'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
  'countries': ['USA'],
  'directors': ['Reginald Barker',
   'Thomas H. Ince',
   'Raymond B. West',
   'Walter Edwards',
   'David Hartford',
   'Jay Hunt',
   'J. Parker Read Jr.'],
  'fullplot': 'Allegorical film about peace. A king starts a war, many of the women are against it, people are pressed into service. A count has constructed a submarine and gets the order to sink an ocean liner, that is also carrying - supposedly - ammunition for the enemy. The count refuses to fire the torpedos, and sinks the submarine. He survives, but in a limbo between death and life where he meets Jesus, who takes him over to preach peace. Naturally the king arrests him and sentences him to death for treason, but then Jesus shows him the real face of war.',
  'genres': ['Drama'],
  'imdb': {'id': 6517, 'rating': 6.3

Jika dibandingkan, dapat dilihat bahwa antara values yang ada pada field  collection `movies` berbeda dengan `clean_movies_devita` yakni pada tomatoes dan num_mflux_comments. Hal ini dikarenakan pada collection `movies_initial` tidak memiliki data tersebut. namun untuk format field antara kedua collection telah sama. 

### Conclusion :

1. **Document yang ada pada `movies` dan `clean_movies_devita` memiliki perbedaan dalam segi jumlah document.**
2. **format field antara kedua collection telah sama namun pada value dari masing-masing field terdapat beberapa perbedaan, hal ini dikarenakan pada `movies_initial` tidak memiliki data yang sama dengan `movies`.**